fish-rust/src/builtins/string.rs

```diff
@@ -26,118 +25,116 @@ macro_rules! string_error {
     $(,)?
     ) => {
         $streams.err.append(L!("string "));
-        $streams.err.append(wgettext_fmt!($string, $($args),*));
+        $streams.err.append(&wgettext_fmt!($string, $($args),*));
     };
 }
 use string_error;
 
-fn string_unknown_option(
-    parser: &mut parser_t,
-    streams: &mut io_streams_t,
-    subcmd: &wstr,
-    opt: &wstr,
-) {
+fn string_unknown_option(parser: &Parser, streams: &mut IoStreams<'_>, subcmd: &wstr, opt: &wstr) {
     string_error!(streams, BUILTIN_ERR_UNKNOWN, subcmd, opt);
-    builtin_print_error_trailer(parser, streams, L!("string"));
+    builtin_print_error_trailer(parser, streams.err, L!("string"));
 }
 
 trait StringSubCommand<'args> {
-    const SHORT_OPTIONS: &'static wstr;
-    const LONG_OPTIONS: &'static [woption<'static>];
+    fn short_options(&self) -> &'static wstr;
+    fn long_options(&self) -> &'static [woption<'static>];
 
     /// Parse and store option specified by the associated short or long option.
     fn parse_opt(
         &mut self,
-        name: &wstr,
+        w: &mut wgetopter_t<'static, 'args>,
         c: char,
-        arg: Option<&'args wstr>,
     ) -> Result<(), StringError>;
 
     fn parse_opts(
         &mut self,
-        args: &mut [&'args wstr],
-        parser: &mut parser_t,
-        streams: &mut io_streams_t,
-    ) -> Result<usize, Option<c_int>> {
-        let cmd = args[0];
-        let mut args_read = Vec::with_capacity(args.len());
-        args_read.extend_from_slice(args);
-
-        let mut w = wgetopter_t::new(Self::SHORT_OPTIONS, Self::LONG_OPTIONS, args);
+        w: &mut wgetopter_t<'static, 'args>,
+        parser: &Parser,
+        streams: &mut IoStreams<'_>,
+    ) -> Result<(), Option<c_int>> {
+        let mut args_read = Vec::with_capacity(w.argv().len());
+        args_read.extend_from_slice(w.argv());
+
         while let Some(c) = w.wgetopt_long() {
             match c {
                 ':' => {
                     streams.err.append(L!("string ")); // clone of string_error
-                    builtin_missing_argument(parser, streams, cmd, args_read[w.woptind - 1], false);
+                    builtin_missing_argument(
+                        parser,
+                        streams,
+                        w.cmd(),
+                        &args_read[w.woptind - 1],
+                        false,
+                    );
                     return Err(STATUS_INVALID_ARGS);
                 }
                 '?' => {
-                    string_unknown_option(parser, streams, cmd, args_read[w.woptind - 1]);
+                    string_unknown_option(parser, streams, w.cmd(), &args_read[w.woptind - 1]);
                     return Err(STATUS_INVALID_ARGS);
                 }
                 c => {
-                    let retval = self.parse_opt(cmd, c, w.woptarg);
```

Comment:
https://github.com/fish-shell/fish-shell/commit/d6b41b33f4c888bbce3cba1e14e3dbf9dadc0afd#r124997793
> I wrote: I don't get why this isn't just a change to `let retval = self.parse_opt(w.cmd(), c, w.woptarg())`. This is very noisy. (even if changed to `WString` those methods should give references, I think)

> @krobelus responded:
> > I don't get why this isn't just a change to `let retval = self.parse_opt(w.cmd(), c, w.woptarg())`. This is very noisy. (even if changed to `WString` those methods should give references, I think)
> 
> yeah that's probably wrong (maybe I thought one of them was a mut reference).
The noisiness we can handle by reorganizing commits later ;)